File: 71.rb
require 'open-uri'
require 'pry'

rules = open('7.txt') {|f| f.read }
.split("\n")
.map{ |r| {
  rule: r,
  container: r.split("contain")[0].gsub(/bag(?:s)?/,"").strip,
  contents: r.split("contain")[1].split(", ").map{|r| r.gsub(/bag(?:s)?/,"").gsub(".","").strip}
}}

#pp rules
@results = []
def rulesCanContainColor?(rules, color, depth=0)
  founds = rules.select{|r| r[:contents].select{|c| c.index(color)}.length > 0}  
  puts "depth: #{depth}, searching for #{color}, found: #{founds.length}, #{founds.map{|f| f[:container]}.join(", ")}"
  @results += founds
  #binding.pry
  founds.each{|f| rulesCanContainColor?(rules, f[:container], depth+=1) } if depth < 100
end

rulesCanContainColor?(rules, "shiny gold")
puts @results.map{|r| r[:container]}.uniq.count

File: 71.js
let rules = []
let results = []
fetch('https://adventofcode.com/2020/day/7/input')
.then(response => {
    response.text().then(input => {
        rules = input.split("\n")
        .filter(r => r != "")
        .map(r => {
            let container = r.split("contain ")[0]
            .replace("bags", "")
            .replace("bag", "").trim();
            let contents = r.split("contain ")[1].replace(".", "")
            .split(", ")
            .map(c => c.replace("bags","").replace("bag","").trim())
            return {
                container: container,
                contents: contents,
                rule: r,
            }
        })

        let countRulesThatCanContain = (color) => {
            let founds = rules.filter(r => r.contents.filter(r => r.indexOf(color) != -1).length > 0)
            results = [...results, ...founds]
            founds.forEach( f => countRulesThatCanContain(f.container))
        };

        countRulesThatCanContain("shiny gold");
        console.log([... new Set(results.map( r => r.container))].length)
    });
});

